pkg/jobstore/boltdb/store.go

```diff
@@ -281,46 +235,64 @@ func (b *BoltJobStore) getJobs(tx *bolt.Tx, query jobstore.JobQuery) ([]model.Jo
 	}
 
 	jobSet := make(map[string]struct{})
+	tagSet := make(map[string]struct{})
+	clientSet := make(map[string]struct{})
 
 	if query.ReturnAll {
-		bkt, err := NewBucketPath(BucketPathJobs).Get(tx, false)
+		bkt, err := NewBucketPath(BucketJobs).Get(tx, false)
 		if err != nil {
 			return nil, err
 		}
 
-		err = bkt.ForEach(func(k []byte, v []byte) error {
-			if v != nil { // If not a bucket
-				jobSet[string(k)] = struct{}{}
-			}
-
+		err = bkt.ForEachBucket(func(k []byte) error {
+			jobSet[string(k)] = struct{}{}
 			return nil
 		})
 		if err != nil {
 			return nil, err
 		}
 	} else {
-		if query.ClientID != "" {
-			ids, err := b.clientsIndex.List(tx, []byte(query.ClientID))
+		for _, tag := range query.IncludeTags {
+			tagLabel := []byte(strings.ToLower(string(tag)))
+			ids, err := b.tagsIndex.List(tx, tagLabel)
 			if err != nil {
 				return nil, err
 			}
 
 			for _, k := range ids {
-				jobSet[string(k)] = struct{}{}
+				tagSet[string(k)] = struct{}{}
 			}
 		}
 
-		for _, tag := range query.IncludeTags {
-			tagLabel := []byte(strings.ToLower(string(tag)))
-			ids, err := b.tagsIndex.List(tx, tagLabel)
+		if query.ClientID != "" {
+			ids, err := b.clientsIndex.List(tx, []byte(query.ClientID))
 			if err != nil {
 				return nil, err
 			}
 
 			for _, k := range ids {
-				jobSet[string(k)] = struct{}{}
+				clientSet[string(k)] = struct{}{}
 			}
 		}
+
+		clientKeys := maps.Keys(clientSet)
+		clientKeysLen := len(clientKeys)
+
+		tagKeys := maps.Keys(tagSet)
+		tagKeysLen := len(tagKeys)
+
+		var jobIDs []string
+		if clientKeysLen > 0 && tagKeysLen > 0 {
+			jobIDs = lo.Intersect(clientKeys, tagKeys)
+		} else if clientKeysLen > 0 && tagKeysLen == 0 { // being explicit
+			jobIDs = clientKeys
+		} else if tagKeysLen > 0 && clientKeysLen == 0 { // being explicit
+			jobIDs = tagKeys
```

Comment:
we shouldn't check if the keys are empty, but if the query specified clientID or IncludeTags. If the user specified a tag that didn't have any matching jobs, then we return all jobs which is not the desired behaviour

Reply:
Actually the expected behaviour is:
- `ReturnAll` should ignore the `clientID` filter and will include jobs for different clients
-`IncludeTags` if specified, will filter out any jobs that don't include the specified tags
- `ExcludeTags` if specified, will filter out any jobs that has the specified tags

The filer names are confusing and we need to revisit them. Since you are away, I can update this code to get your PR merged.